### PR TITLE
[ENH] Plot histogram for a user-selected column

### LINKED ISSUE #82: Implement plotting of histogram for a user-selected column
- For a categorical column, the plot should show #'s of occurrences of each unique value
- For a continuous column, the plot should just show the distribution

## Steps to implement
- [x] Add dropdown for user to select a column in the datatable to explore/visualize
  - [x] Should only appear if user has uploaded a phenotypic file (and disappear if user selects an imaging file)
  - [x] Should be populated with column names from the phenotypic file (aside from unique identifier columns?)
- [ ] ~Add utility function to determine whether selected column is continuous/numerical or categorical~ EDIT: not necessary, pandas + plotly are smart enough to figure this out
- [x] Add plotting function to generate a histogram based on the column
  - Plotting function should probably take as params (1) the dataframe (2) column name to plot ~and (3) a parameter about the type of column it is (categorical vs. continuous)~
- [x] Make new histogram responsive to datatable filtering
  - [x] If active datatable contains no matches, plot should be empty regardless of column selected but still show the appropriate column name as the x-axis title

See https://plotly.com/python/histograms/

### PR DESCRIPTION
**Pull request by alyssadai**
Closes #82

Changes proposed in this pull request:

- Add dropdown for user to select a datatable column to plot when input is phenotypic
- Add graph component and plotting function for phenotypic column histogram
- Add helper function to wrap very long string values in columns to avoid long histogram tick labels
- Make histogram responsive to datatable filtering
- Update tests:
  - Smoke test for a valid phenotypic bagel now looks for presence of the phenotypic col dropdown
  - Add test that a relevant histogram is displayed when a column is selected from the dropdown to plot
- Remove unused extra `None` return value for callback, which was missed in #71 (not sure why this wasn't causing any errors??)
- More informative comments added

## Checklist

- [x] PR has an interpretable title with a prefix (`[ENH]`, `[FIX]`, `[REF]`, `[TST]`, `[CI]`, `[MNT]`, `[INF]`, `[MODEL]`, `[DOC]`) _(see https://neurobagel.org/contributing/pull_requests for more info)_
- [x] PR links to GitHub issue with mention `Closes #XXXX`
- [x] Tests pass
- [x] Checks pass

For new features:
- [x] Tests have been added

For bug fixes:
- [ ] There is at least one test that would fail under the original bug conditions.

### CODE REVIEW

**Comment by nikhil153** [comment]
Thanks @alyssadai! 

Overall it works great! Minor changes would be:

1. Make histogram style to have separate bars with spaces between to clearly represent bin centers. 
image
2. This might be related to screen sizes / browser, but when first loaded, I could not see the `Select a column to plot:` title until I scrolled down. If it's easy, we should rearrange the layout, so that
- Plotting functionality is seen right away 
- Plotting title has bigger font and should say what it is getting plotted (e.g. count distribution) 
- Plot itself need not cover the entire width of the screen. 
3. Also if it's simple enough, we should add along with the plot
- counts for missing values
- `mean` and `std-dev` for the available (i.e. plotted) values.

**Comment by alyssadai** [comment]
Thanks for the prompt feedback @nikhil153!

1. Make histogram style to have separate bars with spaces between to clearly represent bin centers.Okay, will add spaces between all bars in histogram regardless of whether the x-values are categorical/continuous 👍
2. 
3.

**Comment by alyssadai** [comment]
Thanks for the prompt feedback @nikhil153!

> 1. Make histogram style to have separate bars with spaces between to clearly represent bin centers.

Okay, will update histogram so that there are spaces between all bars regardless of whether the x-values are categorical/continuous 👍

> 2. This might be related to screen sizes / browser, but when first loaded, I could not see the Select a column to plot: title until I scrolled down.

Yeah, this will unfortunately vary according to screen and browser size + zoom so is challenging to standardize. What I can do is
- Make the plotting title font bigger (and also make the title itself more informative as suggested)
- Try and move the column plotting dropdown onto the same row as the "Advanced filtering options" dropdown (this is not currently the arrangement because of the varying numbers of dropdowns for "Advanced filtering options" between imaging and phenotypic dashboard layouts)
- ^if this doesn't work well, I'll try to at least move the column plotting dropdown and plot onto the same row so each is taking up less width-space

Note that even with the above changes, with a narrow/small enough browser elements will still inevitably wrap to the next row.

> 3. Also if it's simple enough, we should add along with the plot
> counts for missing values
> mean and std-dev for the available (i.e. plotted) values.

I'll add these separately as part of addressing https://github.com/neurobagel/proc_dash/issues/96 to make the PR code reviews easier 🙂

**Comment by alyssadai** [comment]
@nikhil153, I've implemented your feedback regarding the histogram plot and dropdown.

Mind quickly testing out the dashboard again to let me know if the updated layout looks good? 🙂 A quick note, I shortened the width of the histogram as suggested, but have left it left-justified for now since a box with the summary stats will be added next to it soon.

**Comment by nikhil153** [comment]
@alyssadai - Looks great! Thanks :)

File: tests/test_app.py [hunk 1]
```diff
@@ -15,22 +15,35 @@ def test_server(dash_duo):
...
+    bagels_path,
+    valid_bagel,
+    bagel_type,
+    expected_elements,
+    unexpected_elements,
 ):
     """
     Smoke test (e2e) for uploading a valid bagel input.
-    Tests that the input filename and an expected element for the uploaded data type is displayed after upload.
+    Tests that the input filename and expected/unexpected elements for the uploaded data type is displayed/not after upload.
```
Comment (surchs) [inline tests/test_app.py:46]: > is displayed/not

I think I get it but my sense is this docstring could be clarified a bit

File: tests/test_app.py [hunk 2]
```diff
@@ -105,3 +125,43 @@ def test_003_upload_invalid_phenotypic_bagel(test_server, bagels_path):
...
+def test_004_phenotypic_col_plotting_selection_generates_histogram(
+    test_server, bagels_path
+):
+    """
+    Given a valid phenotypic bagel, displays a dropdown that, in response to a column option selection,
+    displays a histogram for that column without errors.
+    """
+    upload = test_server.driver.find_element(
+        "xpath",
+        """//*[contains(@id,'"index":"phenotypic","type":"upload-data"')]/div/input""",
```
Comment (surchs) [inline tests/test_app.py:139]: :eyes:

File: tests/test_app.py [hunk 3]
```diff
@@ -105,3 +125,43 @@ def test_003_upload_invalid_phenotypic_bagel(test_server, bagels_path):
     assert (
         test_server.get_logs() == []
     ), "browser console should contain no error"
+
+
+def test_004_phenotypic_col_plotting_selection_generates_histogram(
```
Comment (surchs) [inline tests/test_app.py:130]: very impressive test!

**Review by surchs — APPROVED** [review]
Very cool additions, and impressive tests! Congrats!

All is well, good to go! :cook:

**Pull request merged**